### ISSUE TITLE
the ICE happened when run cargo clippy

### ISSUE THREAD

**Issue by Sherlock-Holo** [rust-lang/rust#93195]
### Code

That's a little hard to find out the minimal reproduce codes but I can give the [project](https://github.com/Sherlock-Holo/rfs) what I am running `cargo clippy`

Alos cargo reports codes is

```
   --> /home/sherlock/.cargo/registry/src/github.com-1ecc6299db9ec823/tonic-0.4.3/src/client/grpc.rs:154:5
    |
154 | /     {
155 | |         let request = if let Some(interceptor) = &self.interceptor {
156 | |             interceptor.call(request)?
157 | |         } else {
...   |
211 | |         Ok(Response::from_http(response))
212 | |     }
    | |_____^


```

### Meta

`rustc --version --verbose`:
```
rustc 1.60.0-nightly (17d29dcdc 2022-01-21)
binary: rustc
commit-hash: 17d29dcdce9b9e838635eb0adefd9b8b1588410b
commit-date: 2022-01-21
host: x86_64-unknown-linux-gnu
release: 1.60.0-nightly
LLVM version: 13.0.0

```

### Error output

```
    Checking tonic v0.4.3
error: internal compiler error: compiler/rustc_mir_transform/src/generator.rs:755:13: Broken MIR: generator contains type http::uri::Parts in MIR, but typeck only knows about {ResumeTy, &mut client::grpc::Grpc<T>, request::Request<S>, C, http::Request<body::BoxBody>, client::grpc::Grpc<T>, <T as GrpcService<body::BoxBody>>::Future, ()} and [&mut client::grpc::Grpc<T>, request::Request<S>, http::uri::PathAndQuery, C]
   --> /home/sherlock/.cargo/registry/src/github.com-1ecc6299db9ec823/tonic-0.4.3/src/client/grpc.rs:154:5
    |
154 | /     {
155 | |         let request = if let Some(interceptor) = &self.interceptor {
156 | |             interceptor.call(request)?
157 | |         } else {
...   |
211 | |         Ok(Response::from_http(response))
212 | |     }
    | |_____^

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/17d29dcdce9b9e838635eb0adefd9b8b1588410b/compiler/rustc_errors/src/lib.rs:1115:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.60.0-nightly (17d29dcdc 2022-01-21) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C link-arg=-fuse-ld=lld --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [optimized_mir] optimizing MIR for `client::grpc::<impl at /home/sherlock/.cargo/registry/src/github.com-1ecc6299db9ec823/tonic-0.4.3/src/client/grpc.rs:35:1: 213:2>::streaming::{closure#0}`
end of query stack
error: could not compile `tonic`
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at 'Box<dyn Any>', /rustc/17d29dcdce9b9e838635eb0adefd9b8b1588410b/compiler/rustc_errors/src/lib.rs:1115:9
stack backtrace:
   0: std::panicking::begin_panic::<rustc_errors::ExplicitBug>
   1: std::panic::panic_any::<rustc_errors::ExplicitBug>
   2: <rustc_errors::HandlerInner>::span_bug::<rustc_span::span_encoding::Span>
   3: <rustc_errors::Handler>::span_bug::<rustc_span::span_encoding::Span>
   4: rustc_middle::ty::context::tls::with_opt::<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, ()>
   5: rustc_middle::util::bug::opt_span_bug_fmt::<rustc_span::span_encoding::Span>
   6: rustc_middle::util::bug::span_bug_fmt::<rustc_span::span_encoding::Span>
   7: <rustc_mir_transform::generator::StateTransform as rustc_middle::mir::MirPass>::run_pass
   8: rustc_mir_transform::pass_manager::run_passes
   9: rustc_mir_transform::optimized_mir
  10: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::DefId, &rustc_middle::mir::Body>>
  11: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::optimized_mir
  12: <rustc_metadata::rmeta::encoder::EncodeContext>::encode_crate_root
  13: rustc_metadata::rmeta::encoder::encode_metadata_impl
  14: rustc_data_structures::sync::join::<rustc_metadata::rmeta::encoder::encode_metadata::{closure#0}, rustc_metadata::rmeta::encoder::encode_metadata::{closure#1}, rustc_metadata::rmeta::encoder::EncodedMetadata, ()>
  15: rustc_metadata::rmeta::encoder::encode_metadata
  16: <rustc_interface::passes::QueryContext>::enter::<<rustc_interface::queries::Queries>::ongoing_codegen::{closure#0}::{closure#0}, core::result::Result<alloc::boxed::Box<dyn core::any::Any>, rustc_errors::ErrorReported>>
  17: <rustc_interface::queries::Queries>::ongoing_codegen
  18: <rustc_interface::interface::Compiler>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}, core::result::Result<core::option::Option<rustc_interface::queries::Linker>, rustc_errors::ErrorReported>>
  19: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorReported>, rustc_interface::interface::create_compiler_and_run<core::result::Result<(), rustc_errors::ErrorReported>, rustc_driver::run_compiler::{closure#1}>::{closure#1}>
  20: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::util::setup_callbacks_and_run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorReported>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorReported>>::{closure#0}::{closure#0}, core::result::Result<(), rustc_errors::ErrorReported>>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.60.0-nightly (17d29dcdc 2022-01-21) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C link-arg=-fuse-ld=lld --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [optimized_mir] optimizing MIR for `client::grpc::<impl at /home/sherlock/.cargo/registry/src/github.com-1ecc6299db9ec823/tonic-0.4.3/src/client/grpc.rs:35:1: 213:2>::streaming::{closure#0}`
end of query stack
error: could not compile `tonic`
```

</p>
</details>

**Comment by lqd** [comment]
This looks like a duplicate of #93161 and will be fixed by #93165 shortly.

**Comment by lqd** [comment]
I've verified elsewhere that building `tonic` does not ICE with #93165 indeed (via [rustup-toolchain-master](https://github.com/kennytm/rustup-toolchain-install-master/) of the PR's `try` artifacts `d634ef421795e0140724aa9fddf8137b83e616c9` ). It's expected that this would manifest when using `clippy` as well, so I'll close this issue as a duplicate, albeit in a different context, of #93161.

Feel free to reopen if doesn't completely fix this use-case, though I'm confident it will.

Thanks for letting us know.
